LG ESS: add battery control

zaj78 opened the pull request:
Fix #16637, replace #18416, #18492
Rebased on 0.200.8, reworked installer login
Required firmware for active battery control on Home 8/10: at least R2155, tested with R2173.

andig committed:
LgEss: add battery control

andig committed:
wip

andig committed:
Update meter/lgess.go

andig committed:
Update meter/lgpcs/lgpcs.go

andig committed:
Update meter/lgpcs/lgpcs.go

uwen70 committed:
LgEss: add battery control (#18492)

zaj78 committed:
LgEss: separate passwords for user and installer login

andig commented on meter/lgpcs/lgpcs.go:
```diff
@@ -125,46 +135,68 @@ func (m *Com) Data() (EssData, error) {
 	}
 
-	req, err := request.New(http.MethodPost, m.uri+MeterURI, request.MarshalJSON(data), request.JSONEncoding)
+	if err := mapstructure.Decode(payload, &data); err != nil {
```
Why use mapstructure here at all?

zaj78 replied:
I guess to put payload's values into data? Is there a better way? It's actually your code from #18416 ;-)

zaj78 replied:
If the payload is always a map itself, maybe declare it like this? I'm a total Golang newbie...

andig commented on meter/lgpcs/lgpcs.go:
```diff
@@ -125,46 +135,68 @@ func (m *Com) Data() (EssData, error) {
+	return m.request(func(body io.ReadSeeker) (*http.Request, error) {
+		uri := fmt.Sprintf("%s/v1/user/setting/batt", m.uri)
+		return request.New(http.MethodPut, uri, body, request.JSONEncoding)
+	}, map[string]string{
+		"backupmode": mode,
+		"backup_soc": strconv.Itoa(soc),
+		"autocharge": strconv.Itoa(cast.ToInt(autocharge)),
```
```suggestion
	}, map[string]any{
		"backupmode": mode,
		"backup_soc": soc,
		"autocharge": cast.ToInt(autocharge),
```

zaj78 replied:
Wird das so nicht ohne Hochkommata ins Request gerendert? Weiß nicht, ob die Anlage das mitmacht, und kann momentan nicht testen, weil Batterie defekt ist :-(

andig replied:
Ah, das muss String sein- ok, dann geht das nicht.

andig commented:
Worth noting that this instance user/pwd comparison is likely broken since it limits to a single device. Instead there should be an instance cache that returns a cached instance.

zaj78 commented:
> Worth noting that this instance user/pwd comparison is likely broken since it limits to a single device. Instead there should be an instance cache that returns a cached instance.

Vermutlich deswegen funktioniert es auch nicht mit den Angaben nur bei einem der Geräte. Wie macht man das mit cached instance? Und sollte das nicht lieber auf eine eigene Baustelle?

andig commented:
Auf jeden Fall eigene Baustelle. Lösung z.b. analog RCT.

andig committed:
Simplify

andig commented:
Ich hab noch ein klein bisschen den Request Mechanismus aufgeräumt und vereinfacht- könntest Du nochmal kurz testen? Dann wäre das hoffentlich good to go.

zaj78 committed:
Amend simplification

- essInfo request needs auth_key;
- check the right err for StatusMethodNotAllowed

andig commented on meter/lgpcs/lgpcs.go:
```diff
@@ -136,9 +136,9 @@ func (m *Com) Data() (EssData, error) {
 // essInfo reads essinfo/home
 func (m *Com) essInfo() (EssData, error) {
-	f := func(_ any) (*http.Request, error) {
+	f := func(payload any) (*http.Request, error) {
```
Wozu? Das ist immer nil.

zaj78 replied:
Nein, da sit auth_key drin

andig replied:
Oh, das ist tricky :O

zaj78 commented:
Ist das Konfig-Bieispiel aus meter/lgess.go absichtlich weg?

andig commented:
> Ist das Konfig-Bieispiel aus meter/lgess.go absichtlich weg?

Ja :)

Sami0178 commented:
Ihr seid großartig!
Wie geht es hier weiter? Kann ich den aktuellen Stand probieren?
Mein Gerät hat folgende FW. (Update steht auf automatisch) 
Geht es damit?
Versions info
- PMS SW Version : 10.05.7394 / 02.02.2024 R1955
- PCS SW Version : LG 05.00.01.00 R180 1.176.3
- BMS SW Version : BMS 01.07.00.01 / DCDC 7K.4.8 
BMS SW Version
- BMS SW Unit1_Version : BMS 01.07.00.01 / DCDC 7K.4.8

marcelGoerentz commented:
@zaj78 Wo finde ich denn die Firmware? Hab aktuell auch nur die R1955.
Würde das update per USB wagen.

zaj78 commented:
@marcelGoerentz: zum Firmware-Update gibt es z.B. [hier](https://github.com/evcc-io/evcc/issues/16637#issuecomment-2558204025) etwas.
Im Code wird die Version m.E. nicht geprüft. Wenn jemand weiß, wie man diese herausbekommt, könnte man das einbauen. Hat man z.B. in ioBrocker oder Home Assistant bereits die Version aus dem Gerät herausgekitzelt?

Sami0178 commented:
@marcelGoerentz @zaj78 
Habe letzte Woche bei LG Support angefragt, leider keine Antwort.
Der oben genannte Link zu R2178 geht bei mir leider nicht. 
Ich werde es mit link zu 2159 aus PV Forum versuchen.
https://ess-pr-web.s3.eu-central-1.amazonaws.com/firmware/P10.05.7433R2159_D5003.0198_USR_G4V_D10k.bin

marcelGoerentz commented:
> @marcelGoerentz @zaj78 
> Habe letzte Woche bei LG Support angefragt, leider keine Antwort.
> Der oben genannte Link zu R2178 geht bei mir leider nicht. 
> Ich werde es mit link zu 2159 aus PV Forum versuchen.
> https://ess-pr-web.s3.eu-central-1.amazonaws.com/firmware/P10.05.7433R2159_D5003.0198_USR_G4V_D10k.bin
> 

Du hast Recht. Sorry 
Hab den Link jetzt korrigiert!!!

marcelGoerentz commented:
> @marcelGoerentz: Im Code wird die Version m.E. nicht geprüft. Wenn jemand weiß, wie man diese herausbekommt, könnte man das einbauen. Hat man z.B. in ioBrocker oder Home Assistant bereits die Version aus dem Gerät herausgekitzelt?

Über Modbus TCP lässt sich die Version auslesen, wenn SunSpec aktiviert wurde. Ich schaue aber gerne Mal nach ob ich die auch per Http-Request abfragen kann. Ich meine in der App wird sie ja auch angezeigt 😉

marcelGoerentz commented:
Guten Morgen.

Hier der Response vom angesprochenen Endpoint:
<details>
  <summary>Response</summary>
  ```json
  {
    "pms": {
      "model": "D008KE1N211",
      "serialno": "DE2012BKRE0160D7",
      "ac_input_power": "13500",
      "ac_output_power": "8",
      "install_date": "2021-10-27"
    },
    "batt": {
      "capacity": "98",
      "type": "hb",
      "nameplate_energy_1": "9800",
      "hbc_cycle_count_1": "0",
      "hbc_remaining_cap_1": "8518",
      "hbc_dischg_rate_1": "0",
      "hbc_chg_energy_1": "2500000000",
      "hbc_dischg_energy_1": "2500000000",
      "hbc_chg_cap_1": "0",
      "hbc_dischg_cap_1": "0",
      "hbc_deep_dischg_cnt_1": "0",
      "hbc_over_chg_cnt_1": "0",
      "nameplate_energy_2": "0",
      "hbc_cycle_count_2": "0",
      "hbc_remaining_cap_2": "0",
      "hbc_dischg_rate_2": "0",
      "hbc_chg_energy_2": "0",
      "hbc_dischg_energy_2": "0",
      "hbc_chg_cap_2": "0",
      "hbc_dischg_cap_2": "0",
      "hbc_deep_dischg_cnt_2": "0",
      "hbc_over_chg_cnt_2": "0",
      "hbc_a_serials": "",
      "hbc_b_serials": "",
      "hbc_a_pack_dates": "",
      "hbc_b_pack_dates": "",
      "install_date": "2020-10-27"
    },
    "version": {
      "pms_version": "10.05.7437",
      "pms_build_date": "2025-01-14 R2173",
      "pcs_version": "LG 05.00.01.00 R201 1.197.3",
      "bms_version": "BMS 01.07.00.01 / DCDC 7K.4.8 ",
      "bms_unit1_version": "BMS 01.07.00.01 / DCDC 7K.4.8 ",
      "bms_unit2_version": " "
    }
  }
  ```
</details>
